Ensure `swap_nonoverlapping` is really always untyped

scottmcm opened the pull request:
This replaces #134954, which was arguably overcomplicated.

## Fixes #134713

Actually using the type passed to `ptr::swap_nonoverlapping` for anything other than its size + align turns out to not work, so this goes back to always erasing the types down to just bytes.

(Except in `const`, which keeps doing the same thing as before to preserve @RalfJung's fix from #134689)

## Fixes #134946

I'd previously moved the swapping to use auto-vectorization *on bytes*, but someone pointed out on Discord that the tail loop handling from that left a whole bunch of byte-by-byte swapping around.  This goes back to manual tail handling to avoid that, then still triggers auto-vectorization on pointer-width values.  (So you'll see `<4 x i64>` on `x86-64-v3` for example.)

scottmcm committed:
Ensure `swap_nonoverlapping` is really always untyped

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13469829668/job/37641959202) [(plain)](https://github.com/rust-lang/rust/commit/60d61a861477a874965355e8a451bc508219227b/checks/37641959202/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 28.0s done
#22 DONE 33.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/missing_span_in_backtrace/missing_span_in_backtrace.stderr"
diff of stderr:

5    |
6 note: inside `std::ptr::read::<MaybeUninit<MaybeUninit<u8>>>`
7   --> $SRC_DIR/core/src/ptr/mod.rs:LL:COL
- note: inside `std::ptr::swap_nonoverlapping_simple_untyped::<MaybeUninit<u8>>`
+ note: inside `std::ptr::swap_nonoverlapping_const::<MaybeUninit<u8>>`
9   --> $SRC_DIR/core/src/ptr/mod.rs:LL:COL
10 note: inside `swap_nonoverlapping::compiletime::<MaybeUninit<u8>>`
11   --> $SRC_DIR/core/src/ptr/mod.rs:LL:COL

The actual stderr differed from the expected stderr.
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args consts/missing_span_in_backtrace.rs`
To only update this specific test, also pass `--test-args consts/missing_span_in_backtrace.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/consts/missing_span_in_backtrace.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/missing_span_in_backtrace" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "ui-testing=no"
--- stderr -------------------------------
error[E0080]: evaluation of constant value failed
##[error]  --> /rustc/FAKE_PREFIX/library/core/src/ptr/mod.rs:1413:9
   |
   |
   = note: unable to copy parts of a pointer from memory at alloc58
   |
note: inside `std::ptr::read::<MaybeUninit<MaybeUninit<u8>>>`
note: inside `std::ptr::swap_nonoverlapping_const::<MaybeUninit<u8>>`
  --> /rustc/FAKE_PREFIX/library/core/src/ptr/mod.rs:1127:37
  --> /rustc/FAKE_PREFIX/library/core/src/ptr/mod.rs:1127:37
note: inside `swap_nonoverlapping::compiletime::<MaybeUninit<u8>>`
  --> /rustc/FAKE_PREFIX/library/core/src/ptr/mod.rs:1098:22
note: inside `swap_nonoverlapping::<MaybeUninit<u8>>`
note: inside `X`
  --> /checkout/tests/ui/consts/missing_span_in_backtrace.rs:16:9
   |
16 | /         ptr::swap_nonoverlapping(
16 | /         ptr::swap_nonoverlapping(
17 | |             &mut ptr1 as *mut _ as *mut MaybeUninit<u8>,
18 | |             &mut ptr2 as *mut _ as *mut MaybeUninit<u8>,
19 | |             mem::size_of::<&i32>(),
   | |_________^
   = help: this code performed an operation that depends on the underlying bytes representing a pointer
   = help: the absolute address of a pointer is not known at compile-time, so such operations are not supported
   = note: this error originates in the macro `$crate::intrinsics::const_eval_select` which comes from the expansion of the macro `const_eval_select` (in Nightly builds, run with -Z macro-backtrace for more info)
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13470073644/job/37642476566) [(plain)](https://github.com/rust-lang/rust/commit/6964320a3ceab6f7dc8d75461fcae122fda66aa9/checks/37642476566/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 28.6s done
#22 DONE 34.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/swap-small-types.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/swap-small-types/swap-small-types.ll" "/checkout/tests/codegen/swap-small-types.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/swap-small-types.rs:90:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: alloca
               ^
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/swap-small-types/swap-small-types.ll:87:17: note: found here
 %_45.i.i.i.i = alloca [48 x i8], align 8

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/swap-small-types/swap-small-types.ll
Check file: /checkout/tests/codegen/swap-small-types.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
        1: ; ModuleID = 'swap_small_types.c15a5a94bf56ed41-cgu.0' 
        2: source_filename = "swap_small_types.c15a5a94bf56ed41-cgu.0" 
        3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
        5:  
        5:  
        6: %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>" = type { [8 x i8] } 
        7:  
        8: @alloc_78efff4de17083504b82e006a866bf7b = private unnamed_addr constant <{ [46 x i8] }> <{ [46 x i8] c"/rustc/FAKE_PREFIX/library/core/src/ptr/mod.rs" }>, align 1 
        9: @alloc_aa11bb599a91239fe9a838c45d54e317 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_78efff4de17083504b82e006a866bf7b, [16 x i8] c".\00\00\00\00\00\00\00\9D\04\00\00\09\00\00\00" }>, align 8 
       10:  
       11: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite) uwtable 
       12: define void @swap_rgb48_manually(ptr noalias nocapture noundef align 2 dereferenceable(6) %x, ptr noalias nocapture noundef align 2 dereferenceable(6) %y) unnamed_addr #0 { 
       13: start: 
       14:  %temp = alloca [6 x i8], align 2 
       15:  call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 2 dereferenceable(6) %temp, ptr noundef nonnull align 2 dereferenceable(6) %x, i64 6, i1 false) 
       16:  tail call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 2 dereferenceable(6) %x, ptr noundef nonnull align 2 dereferenceable(6) %y, i64 6, i1 false) 
       17:  call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 2 dereferenceable(6) %y, ptr noundef nonnull align 2 dereferenceable(6) %temp, i64 6, i1 false) 
       18:  ret void 
       19: } 
       20:  
       21: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite, inaccessiblemem: readwrite) uwtable 
       22: define void @swap_rgb48(ptr noalias nocapture noundef align 2 dereferenceable(6) %x, ptr noalias nocapture noundef align 2 dereferenceable(6) %y) unnamed_addr #1 personality ptr @rust_eh_personality { 
       23: start: 
       24:  tail call void @llvm.experimental.noalias.scope.decl(metadata !3) 
       25:  tail call void @llvm.experimental.noalias.scope.decl(metadata !6) 
       26:  %_3.sroa.0.0.copyload.i.i6.i.i.i = load i32, ptr %x, align 2, !alias.scope !3, !noalias !6 
       27:  %_4.sroa.0.0.copyload.i.i7.i.i.i = load i32, ptr %y, align 2, !alias.scope !6, !noalias !3 
       28:  store i32 %_4.sroa.0.0.copyload.i.i7.i.i.i, ptr %x, align 2, !alias.scope !3, !noalias !6 
       29:  store i32 %_3.sroa.0.0.copyload.i.i6.i.i.i, ptr %y, align 2, !alias.scope !6, !noalias !3 
       30:  %self7.i.i.i.i = getelementptr inbounds i8, ptr %x, i64 4 
       31:  %self8.i.i.i.i = getelementptr inbounds i8, ptr %y, i64 4 
       32:  tail call void @llvm.experimental.noalias.scope.decl(metadata !8) 
       33:  tail call void @llvm.experimental.noalias.scope.decl(metadata !11) 
       34:  %_3.sroa.0.0.copyload.i12.i.i.i.i = load i16, ptr %self7.i.i.i.i, align 2, !alias.scope !8, !noalias !11 
       35:  %_4.sroa.0.0.copyload.i13.i.i.i.i = load i16, ptr %self8.i.i.i.i, align 2, !alias.scope !11, !noalias !8 
       36:  store i16 %_4.sroa.0.0.copyload.i13.i.i.i.i, ptr %self7.i.i.i.i, align 2, !alias.scope !8, !noalias !11 
       37:  store i16 %_3.sroa.0.0.copyload.i12.i.i.i.i, ptr %self8.i.i.i.i, align 2, !alias.scope !11, !noalias !8 
       38:  ret void 
       39: } 
       40:  
       41: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite, inaccessiblemem: readwrite) uwtable 
       42: define void @swap_rgba64(ptr noalias nocapture noundef align 2 dereferenceable(8) %x, ptr noalias nocapture noundef align 2 dereferenceable(8) %y) unnamed_addr #1 personality ptr @rust_eh_personality { 
       43: start: 
       44:  tail call void @llvm.experimental.noalias.scope.decl(metadata !13) 
       45:  tail call void @llvm.experimental.noalias.scope.decl(metadata !16) 
       46:  %_3.sroa.0.0.copyload.i.i.i.i.i = load i64, ptr %x, align 2, !alias.scope !13, !noalias !16 
       47:  %_4.sroa.0.0.copyload.i.i.i.i.i = load i64, ptr %y, align 2, !alias.scope !16, !noalias !13 
       48:  store i64 %_4.sroa.0.0.copyload.i.i.i.i.i, ptr %x, align 2, !alias.scope !13, !noalias !16 
       49:  store i64 %_3.sroa.0.0.copyload.i.i.i.i.i, ptr %y, align 2, !alias.scope !16, !noalias !13 
       50:  ret void 
       51: } 
       52:  
       53: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite, inaccessiblemem: readwrite) uwtable 
       54: define void @swap_vecs(ptr noalias nocapture noundef align 8 dereferenceable(24) %x, ptr noalias nocapture noundef align 8 dereferenceable(24) %y) unnamed_addr #1 personality ptr @rust_eh_personality { 
       55: start: 
       56:  %0 = load <2 x i64>, ptr %y, align 8, !alias.scope !18, !noalias !22 
       57:  %1 = load <2 x i64>, ptr %x, align 8, !alias.scope !23, !noalias !22 
       58:  store <2 x i64> %0, ptr %x, align 8, !alias.scope !23, !noalias !22 
       59:  store <2 x i64> %1, ptr %y, align 8, !alias.scope !18, !noalias !22 
       60:  %_12.i.i.2.i.i = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>", ptr %x, i64 2 
       61:  %_14.i.i.2.i.i = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>", ptr %y, i64 2 
       62:  tail call void @llvm.experimental.noalias.scope.decl(metadata !26) 
       63:  tail call void @llvm.experimental.noalias.scope.decl(metadata !28) 
       64:  %_3.sroa.0.0.copyload.i.i.i.2.i.i = load i64, ptr %_12.i.i.2.i.i, align 8, !alias.scope !26, !noalias !28 
       65:  %_4.sroa.0.0.copyload.i.i.i.2.i.i = load i64, ptr %_14.i.i.2.i.i, align 8, !alias.scope !28, !noalias !26 
       66:  store i64 %_4.sroa.0.0.copyload.i.i.i.2.i.i, ptr %_12.i.i.2.i.i, align 8, !alias.scope !26, !noalias !28 
       67:  store i64 %_3.sroa.0.0.copyload.i.i.i.2.i.i, ptr %_14.i.i.2.i.i, align 8, !alias.scope !28, !noalias !26 
       68:  ret void 
       69: } 
       70:  
       71: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: readwrite) uwtable 
       72: define void @swap_slices(ptr noalias nocapture noundef align 8 dereferenceable(16) %x, ptr noalias nocapture noundef align 8 dereferenceable(16) %y) unnamed_addr #0 { 
       73: start: 
       74:  %0 = load ptr, ptr %x, align 8, !nonnull !22, !align !30, !noundef !22 
       75:  %1 = getelementptr inbounds i8, ptr %x, i64 8 
       76:  %2 = load i64, ptr %1, align 8, !noundef !22 
       77:  tail call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 8 dereferenceable(16) %x, ptr noundef nonnull align 8 dereferenceable(16) %y, i64 16, i1 false) 
       78:  store ptr %0, ptr %y, align 8 
       79:  %3 = getelementptr inbounds i8, ptr %y, i64 8 
       80:  store i64 %2, ptr %3, align 8 
       81:  ret void 
       82: } 
       83:  
       84: ; Function Attrs: nonlazybind uwtable 
       85: define void @swap_rgb24_slices(ptr noalias nocapture noundef nonnull align 1 %x.0, i64 noundef %x.1, ptr noalias nocapture noundef nonnull align 1 %y.0, i64 noundef %y.1) unnamed_addr #2 { 
       87:  %_45.i.i.i.i = alloca [48 x i8], align 8 
       87:  %_45.i.i.i.i = alloca [48 x i8], align 8 
not:90                     !~~~~~                     error: no match expected
       88:  %i.i.i.i.i = alloca [8 x i8], align 8 
       89:  %bytes1.i.i.i.i = alloca [8 x i8], align 8 
       90:  %_3 = icmp eq i64 %x.1, %y.1 
       91:  br i1 %_3, label %bb1.i, label %bb3 
       92:  
       93: bb1.i: ; preds = %start 
       94:  tail call void @llvm.experimental.noalias.scope.decl(metadata !31) 
       95:  tail call void @llvm.experimental.noalias.scope.decl(metadata !34) 
       96:  %0 = icmp eq i64 %x.1, 0 
       97:  br i1 %0, label %bb3, label %bb4.i.i 
       98:  
       99: bb4.i.i: ; preds = %bb1.i 
      100:  %bytes.i.i = mul nuw i64 %x.1, 3 
      101:  %tail.i.i.i = and i64 %bytes.i.i, 7 
      102:  %1 = icmp ult i64 %x.1, 3 
      103:  br i1 %1, label %bb3.i.i.i, label %bb1.i.i.i 
      104:  
      105: bb1.i.i.i: ; preds = %bb4.i.i 
      106:  %chunks5.i.i.i = lshr i64 %bytes.i.i, 3 
      107:  %min.iters.check = icmp ult i64 %x.1, 11 
      108:  br i1 %min.iters.check, label %bb3.i.i.i.i.preheader, label %vector.ph 
      109:  
      110: vector.ph: ; preds = %bb1.i.i.i 
      111:  %n.vec = and i64 %chunks5.i.i.i, 2305843009213693948 
      112:  br label %vector.body 
      113:  
      114: vector.body: ; preds = %vector.body, %vector.ph 
      115:  %index = phi i64 [ 0, %vector.ph ], [ %index.next, %vector.body ] 
      116:  %2 = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>", ptr %x.0, i64 %index 
      117:  %3 = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>", ptr %y.0, i64 %index 
      118:  tail call void @llvm.experimental.noalias.scope.decl(metadata !36) 
      119:  tail call void @llvm.experimental.noalias.scope.decl(metadata !39) 
      120:  %4 = getelementptr inbounds i64, ptr %2, i64 2 
      121:  %wide.load = load <2 x i64>, ptr %2, align 1, !alias.scope !41, !noalias !42 
      122:  %wide.load1 = load <2 x i64>, ptr %4, align 1, !alias.scope !41, !noalias !42 
      123:  %5 = getelementptr inbounds i64, ptr %3, i64 2 
      124:  %wide.load2 = load <2 x i64>, ptr %3, align 1, !alias.scope !42, !noalias !41 
      125:  %wide.load3 = load <2 x i64>, ptr %5, align 1, !alias.scope !42, !noalias !41 
      126:  store <2 x i64> %wide.load2, ptr %2, align 1, !alias.scope !41, !noalias !42 
      127:  store <2 x i64> %wide.load3, ptr %4, align 1, !alias.scope !41, !noalias !42 
      128:  store <2 x i64> %wide.load, ptr %3, align 1, !alias.scope !42, !noalias !41 
      129:  store <2 x i64> %wide.load1, ptr %5, align 1, !alias.scope !42, !noalias !41 
      130:  %index.next = add nuw i64 %index, 4 
      131:  %6 = icmp eq i64 %index.next, %n.vec 
      132:  br i1 %6, label %middle.block, label %vector.body, !llvm.loop !43 
      133:  
      134: middle.block: ; preds = %vector.body 
      135:  %cmp.n = icmp eq i64 %chunks5.i.i.i, %n.vec 
      136:  br i1 %cmp.n, label %bb3.i.i.i, label %bb3.i.i.i.i.preheader 
      137:  
      138: bb3.i.i.i.i.preheader: ; preds = %bb1.i.i.i, %middle.block 
      139:  %iter.sroa.0.03.i.i.i.i.ph = phi i64 [ 0, %bb1.i.i.i ], [ %n.vec, %middle.block ] 
      140:  br label %bb3.i.i.i.i 
      141:  
      142: bb3.i.i.i.i: ; preds = %bb3.i.i.i.i.preheader, %bb3.i.i.i.i 
      143:  %iter.sroa.0.03.i.i.i.i = phi i64 [ %_0.i.i.i.i.i, %bb3.i.i.i.i ], [ %iter.sroa.0.03.i.i.i.i.ph, %bb3.i.i.i.i.preheader ] 
      144:  %_0.i.i.i.i.i = add nuw nsw i64 %iter.sroa.0.03.i.i.i.i, 1 
      145:  %_12.i.i.i.i = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>", ptr %x.0, i64 %iter.sroa.0.03.i.i.i.i 
      146:  %_14.i.i.i.i = getelementptr inbounds %"core::mem::maybe_uninit::MaybeUninit<[u8; 8]>", ptr %y.0, i64 %iter.sroa.0.03.i.i.i.i 
      147:  tail call void @llvm.experimental.noalias.scope.decl(metadata !36) 
      148:  tail call void @llvm.experimental.noalias.scope.decl(metadata !39) 
      149:  %_3.sroa.0.0.copyload.i.i.i.i.i = load i64, ptr %_12.i.i.i.i, align 1, !alias.scope !41, !noalias !42 
      150:  %_4.sroa.0.0.copyload.i.i.i.i.i = load i64, ptr %_14.i.i.i.i, align 1, !alias.scope !42, !noalias !41 
      151:  store i64 %_4.sroa.0.0.copyload.i.i.i.i.i, ptr %_12.i.i.i.i, align 1, !alias.scope !41, !noalias !42 
      152:  store i64 %_3.sroa.0.0.copyload.i.i.i.i.i, ptr %_14.i.i.i.i, align 1, !alias.scope !42, !noalias !41 
      153:  %exitcond.not.i.i.i.i = icmp eq i64 %_0.i.i.i.i.i, %chunks5.i.i.i 
      154:  br i1 %exitcond.not.i.i.i.i, label %bb3.i.i.i, label %bb3.i.i.i.i, !llvm.loop !46 
      155:  
      156: bb3.i.i.i: ; preds = %bb3.i.i.i.i, %middle.block, %bb4.i.i 
      157:  %7 = icmp eq i64 %tail.i.i.i, 0 
      158:  br i1 %7, label %bb3, label %bb4.i.i.i 
      159:  
      160: bb4.i.i.i: ; preds = %bb3.i.i.i 
      161:  %8 = and i64 %bytes.i.i, -8 
      162:  %9 = getelementptr inbounds i8, ptr %x.0, i64 %8 
      163:  %_20.i.i.i = getelementptr inbounds i8, ptr %y.0, i64 %8 
      164:  call void @llvm.lifetime.start.p0(i64 8, ptr nonnull %bytes1.i.i.i.i), !noalias !47 
      165:  store i64 %tail.i.i.i, ptr %bytes1.i.i.i.i, align 8, !noalias !47 
      166:  call void @llvm.lifetime.start.p0(i64 8, ptr nonnull %i.i.i.i.i), !noalias !47 
      167:  store i64 0, ptr %i.i.i.i.i, align 8, !noalias !47 
      168:  %_6.i.i.i.i = and i64 %bytes.i.i, 4 
      169:  %10 = icmp eq i64 %_6.i.i.i.i, 0 
      170:  br i1 %10, label %bb4.i.i.i.i, label %bb1.i.i.i.i 
      171:  
      172: bb1.i.i.i.i: ; preds = %bb4.i.i.i 
      173:  tail call void @llvm.experimental.noalias.scope.decl(metadata !48) 
      174:  tail call void @llvm.experimental.noalias.scope.decl(metadata !51) 
      175:  %_3.sroa.0.0.copyload.i.i6.i.i.i = load i32, ptr %9, align 1, !alias.scope !53, !noalias !54 
      176:  %_4.sroa.0.0.copyload.i.i7.i.i.i = load i32, ptr %_20.i.i.i, align 1, !alias.scope !54, !noalias !53 
      177:  store i32 %_4.sroa.0.0.copyload.i.i7.i.i.i, ptr %9, align 1, !alias.scope !53, !noalias !54 
      178:  store i32 %_3.sroa.0.0.copyload.i.i6.i.i.i, ptr %_20.i.i.i, align 1, !alias.scope !54, !noalias !53 
      179:  store i64 4, ptr %i.i.i.i.i, align 8, !noalias !47 
      180:  br label %bb4.i.i.i.i 
      181:  
      182: bb4.i.i.i.i: ; preds = %bb1.i.i.i.i, %bb4.i.i.i 
      183:  %count6.i.i.i.i = phi i64 [ 0, %bb4.i.i.i ], [ 4, %bb1.i.i.i.i ] 
      184:  %_17.i.i.i.i = and i64 %bytes.i.i, 2 
      185:  %11 = icmp eq i64 %_17.i.i.i.i, 0 
      186:  br i1 %11, label %bb8.i.i.i.i, label %bb5.i.i.i.i 
        .
        .
        .
>>>>>>
```

</details>

RalfJung commented on library/core/src/ptr/mod.rs:
Could you add a Miri test for the example from https://github.com/rust-lang/rust/issues/134713?

scottmcm committed:
Add a Miri test for 134713

RalfJung commented on src/tools/miri/tests/pass/stdlib/swap_nonoverlapping_is_untyped.rs:
That's a new folder, isn't it? We have tons of stdlib tests so this only makes things more confusing I think.

Since this tests a particular issue, what about putting it in `issues/issue-134713-swap_nonoverlapping_untyped.rs` or something like that?

RalfJung replied:
Thanks!

ibraheemdev commented:
Passing this one along because I'm not the best person to review this. r? libs

jhpratt commented:
I'm mostly sticking to the trivial PRs at the moment as my review capacity is limited. Re-rolling again.

r? libs

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -1094,49 +1095,26 @@ pub const unsafe fn swap_nonoverlapping<T>(x: *mut T, y: *mut T, count: usize) {
-/// `swap_nonoverlapping` tries to use) so no need to manually SIMD it.
 #[inline]
-const unsafe fn swap_nonoverlapping_simple_untyped<T>(x: *mut T, y: *mut T, count: usize) {
+const unsafe fn swap_nonoverlapping_const<T>(x: *mut T, y: *mut T, count: usize) {
```
This one still does copies which are "too typed", doesn't it? It copeis at `MaybeUninit<T>` which has the padding problem.

Please make do the copies using `copy_nonoverlapping` instead.

RalfJung replied:
CTFE fails to detect this UB. But that doesn't make it not-UB.

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -1094,49 +1095,26 @@ pub const unsafe fn swap_nonoverlapping<T>(x: *mut T, y: *mut T, count: usize) {
+            let slice = slice_from_raw_parts_mut(x, count);
+            // SAFETY: We have two non-overlapping ranges in memory and they're both
+            // readable so don't include the null address, thus they're at most
+            // `(usize::MAX - 1)/2 = isize::MAX` bytes long.
```
I don't understand this comment. Every readable contiguous range of memory is at most `isize::MAX` bytes long as that is the largest possible allocation.

RalfJung replied:
This is documented at https://doc.rust-lang.org/nightly/std/ptr/index.html#allocated-object

RalfJung reviewed (approved):
The code changes LGTM for the non-`const` path; see the comment for the `const` path.

I did not look at the tests; I think that needs an LLVM/codegen expert. @nikic maybe?

RalfJung commented:
To be clear I only reviewed for correctness, I cannot judge whether all the tricks to get better performance actually work.

cuviper commented on tests/codegen/swap-large-types.rs:
```diff
@@ -52,39 +49,61 @@ pub fn swap_slice(x: &mut [KeccakBuffer], y: &mut [KeccakBuffer]) {
 // CHECK-LABEL: @swap_1kb_slices
 #[no_mangle]
 pub fn swap_1kb_slices(x: &mut [OneKilobyteBuffer], y: &mut [OneKilobyteBuffer]) {
     // CHECK-NOT: alloca
```
Doesn't that goal deserve an inline comment?

bors commented:
:umbrella: The latest upstream changes (presumably #137848) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13714383372/job/38356574911) [(plain)](https://github.com/rust-lang/rust/commit/a6de5577f3f78b3eb9ffc20f66a5a85279f8fb29/checks/38356574911/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     |
note: inside `ptr::ptr_swap_nonoverlapping_is_untyped_inner`
    --> library/coretests/tests/ptr.rs:1013:13
     |
1013 |     assert!(b1[1] == buf2[1]);
     |             ^^^^^ the failure occurred here
note: erroneous constant encountered
    --> library/coretests/tests/ptr.rs:1021:5
     |
1021 |     const { ptr_swap_nonoverlapping_is_untyped_inner() };
---
     |
1021 |     const { ptr_swap_nonoverlapping_is_untyped_inner() };
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0080`.
error: test failed, to rerun pass `-p coretests --test coretests`
error: test failed, to rerun pass `-p coretests --test coretests`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/cargo-miri runner /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/miri/x86_64-unknown-linux-gnu/debug/deps/coretests-7ee443223f26c677 notest --quiet -Z unstable-options --format json` (exit status: 1)
note: test exited abnormally; to see the full output pass --nocapture to the harness.
  local time: Fri Mar  7 06:12:52 UTC 2025
  network time: Fri, 07 Mar 2025 06:12:52 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #138155) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cuviper reviewed (approved):
LGTM!

cuviper commented:
@bors r+

bors commented:
:pushpin: Commit 9648340da87ce0c7a04f93d49147c9fedda912c8 has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 9648340da87ce0c7a04f93d49147c9fedda912c8 -->
<!-- homu: {"type":"Approved","sha":"9648340da87ce0c7a04f93d49147c9fedda912c8","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/pull/139571#issuecomment-2789598180

bors commented:
:hourglass: Trying commit b06a88ff61acb08dd4f7502908485a3eaa74d0a6 with merge bef076dd1a3fd383b67cca18c48f4aaf7157d0ea...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b06a88ff61acb08dd4f7502908485a3eaa74d0a6","merge_sha":"bef076dd1a3fd383b67cca18c48f4aaf7157d0ea"} -->

scottmcm committed:
skip `tests/codegen/swap-small-types` when debug assertions are on

In `swap_nonoverlapping_short` there's a new `debug_assert!`, and if that's enabled then the `alloca`s don't optimize out.

bors commented:
:pushpin: Commit 63dcac8423750369ee251e2bad0945ad765eaf5c has been approved by `cuviper`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cuviper 63dcac8423750369ee251e2bad0945ad765eaf5c -->
<!-- homu: {"type":"Approved","sha":"63dcac8423750369ee251e2bad0945ad765eaf5c","approver":"cuviper","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14389535994/job/40361330973)
Approved by: cuviper
Pushing 0fe8f3454dbe9dda52a254991347e96bec579a6f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cuviper","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14389535994/job/40361330973"},"merge_sha":"0fe8f3454dbe9dda52a254991347e96bec579a6f"} -->

rust-timer commented:
Finished benchmarking commit (0fe8f3454dbe9dda52a254991347e96bec579a6f): [comparison URL](https://perf.rust-lang.org/compare.html?start=2205455d444f5f9da3a52e50f43c0306480218fb&end=0fe8f3454dbe9dda52a254991347e96bec579a6f&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2205455d444f5f9da3a52e50f43c0306480218fb&end=0fe8f3454dbe9dda52a254991347e96bec579a6f&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.4%, 1.7%]   | 5     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-1.1%, -0.3%] | 9     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.2% | [-1.1%, 1.7%]  | 14    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2205455d444f5f9da3a52e50f43c0306480218fb&end=0fe8f3454dbe9dda52a254991347e96bec579a6f&stat=max-rss)
<details>
<summary>Results (primary 1.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.7%  | [2.6%, 5.9%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.5% | [-2.6%, -2.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.2%  | [-2.6%, 5.9%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2205455d444f5f9da3a52e50f43c0306480218fb&end=0fe8f3454dbe9dda52a254991347e96bec579a6f&stat=cycles:u)
<details>
<summary>Results (primary 0.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.7%, 0.7%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.7%, 0.7%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=2205455d444f5f9da3a52e50f43c0306480218fb&end=0fe8f3454dbe9dda52a254991347e96bec579a6f&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 0.8%]   | 17    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.3%, -0.0%] | 37    |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 3     |
| All ❌✅ (primary)                 | 0.0%  | [-0.3%, 0.8%]  | 54    |
</details>

**Bootstrap**: 783.243s -> 781.425s (-0.23%)
**Artifact size**: 366.21 MiB -> 365.92 MiB (-0.08%)

Mark-Simulacrum commented:
Correctness fix, relatively small set of regressed scenarios, compared to number of improvements. Marking as triaged.
